File: src/pcapack/src/sumstats/variances.f90
! This Source Code Form is subject to the terms of the Mozilla Public
! License, v. 2.0. If a copy of the MPL was not distributed with this
! file, You can obtain one at http://mozilla.org/MPL/2.0/.

! Copyright 2013-2014, Schmidt


module variances
  implicit none
  
  
  contains
  
  subroutine variance(n, x, mn, var)
    ! in/out
    integer :: n
    double precision :: x( * ), mn, var
    ! local
    integer :: i
    double precision :: dt
    
    mn = 0
    var = 0
    
    do i = 1, n, 1
      dt = x(i) - mn
      mn = mn + dt/i
      var = var + dt*(x(i) - mn)
    end do
    
    var = var/(n-1)
    
    return
  end subroutine
  
  
  
  ! standard deviation
  subroutine standard_deviation(n, x, mn, sd)
    ! in/out
    integer :: n
    double precision :: x( * ), mn, sd
    ! subroutines
    external            dvar
    
    call dvar(n, x, mn, sd)
    sd = sqrt(sd)
    
    return
  end subroutine
  
  
  
  ! rowvars
  subroutine row_variances(m, n, x, ret)
    ! in/out
    integer :: m, n
    double precision :: x(m, n), ret( * )
    ! local
    integer :: i
    double precision :: mn, var
    ! subroutines
    external            dvar
    
    do i = 1, m, 1
      call dvar(n, x(i, :), mn, var)
      ret(i) = var
    end do
    
    return
  end subroutine
  
  
  
  ! colvars
  subroutine col_variances(m, n, x, ret)
    ! in/out
    integer :: m, n
    double precision :: x(m, n), ret( * )
    ! local
    integer :: i
    double precision :: mn, var
    ! subroutines
    external            dvar
    
    do i = 1, n, 1
      call dvar(n, x(:, i), mn, var)
      ret(i) = var
    end do
    
    return
  end subroutine
  
  
  
  ! rowvars
  subroutine row_standard_deviations(m, n, x, ret)
    ! in/out
    integer :: m, n
    double precision :: x(m, n), ret( * )
    ! local
    integer :: i
    double precision :: mn, var
    ! subroutines
    external            dvar
    
    do i = 1, m, 1
      call dvar(n, x(i, :), mn, var)
      ret(i) = sqrt(var)
    end do
    
    return
  end subroutine
  
  
  
  ! colvars
  subroutine col_standard_deviations(m, n, x, ret)
    ! in/out
    integer :: m, n
    double precision :: x(m, n), ret( * )
    ! local
    integer :: i
    double precision :: mn, var
    ! subroutines
    external            dvar
    
    do i = 1, n, 1
      call dvar(n, x(:, i), mn, var)
      ret(i) = sqrt(var)
    end do
    
    return
  end subroutine
  
  
end module

File: src/pcapack/src/sumstats/variances.c
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

// Copyright 2015, Schmidt

#include <math.h>

double pcapack_variance(const int n, double *x)
{
  int i;
  double mean, var, dt;
  
  mean = 0;
  var = 0;
  
  for (i=0; i<n; i++)
  {
    dt = x[i] - mean;
    mean += dt / ((double) i);
    var += dt*(x[i] - mean);
  }
  
  var /= ((double) n-1);
  
  return var;
}



double pcapack_sdev(const int n, double *x)
{
  double sdev = pcapack_variance(n, x);
  
  return sqrt(sdev);
}



int pcapack_rowvars(const int m, const int n, double *x, double *rowvars)
{
  int i, j;
  
  
}



int pcapack_colvars(const int m, const int n, double *x, double *colvars)
{
  int i, j;
  
  for (j=0; j<n; j++)
    colvars[j] = pcapack_variance(m, x+(m*j));
  
  return 0;
}

// sdevs



int pcapack_colsdev(const int m, const int n, double *x, double *colsdev)
{
  int i, j;
  
  for (j=0; j<n; j++)
    colsdev[j] = pcapack_sdev(m, x+(m*j));
  
  return 0;
}

